### PR TITLE
Add ntohs/htons and ntohl/htonl

### PR DESCRIPTION
**Pull request by colinleroy**

### CODE REVIEW

**Comment by mrdudz** [comment]
32 bit values are passed in A/X/sreg/sreg+1

I usually just make an empty function in C with the signature i want, and then compile it and have a look :)

**Comment by colinleroy** [comment]
> I usually just make an empty function in C with the signature i want, and then compile it and have a look :)

Thank you, that's exactly what I did :)

**Comment by oliverschmidt** [comment]
* The prototypes belong into arpa/inet.h
* The two implementations should go into separate files. There's no point in having users of the 'short' variant link all the code of the 'long' variant.
* I think it would be more elegant to use linker aliases than C macros for "the other way round". See https://github.com/cc65/cc65/blob/master/libsrc/apple2/libref.s for examples.

**Comment by colinleroy** [comment]
Thanks, I addressed your comments. Also I stopped brainfarting and made the long swap much cleaner/faster 🙂

File: doc/funcref.sgml [hunk 1]
```diff
@@ -4388,6 +4398,45 @@ to undefined behaviour.
 </descrip>
 </quote>
 
+<sect1>htonl<label id="htonl"><p>
+
+<quote>
+<descrip>
+<tag/Function/Swaps byte order in a 32 bit word.
+<tag/Header/<tt/<ref id="stdlib.h" name="stdlib.h">/
```
Comment (oliverschmidt) [inline doc/funcref.sgml:4406]: Missing header adjustment.

File: doc/funcref.sgml [hunk 2]
```diff
@@ -4388,6 +4398,45 @@ to undefined behaviour.
...
+</descrip>
+</quote>
+
+
+<sect1>htons<label id="htons"><p>
+
+<quote>
+<descrip>
+<tag/Function/Swaps byte order in a 16 bit word.
+<tag/Header/<tt/<ref id="stdlib.h" name="stdlib.h">/
```
Comment (oliverschmidt) [inline doc/funcref.sgml:4426]: Missing header adjustment.

File: doc/funcref.sgml [hunk 3]
```diff
@@ -5757,6 +5806,44 @@ memory allocated for the driver.
 </descrip>
 </quote>
 
+<sect1>ntohl<label id="ntohl"><p>
+
+<quote>
+<descrip>
+<tag/Function/Swaps byte order in a 32 bit word.
+<tag/Header/<tt/<ref id="stdlib.h" name="stdlib.h">/
```
Comment (oliverschmidt) [inline doc/funcref.sgml:5814]: Missing header adjustment.

File: doc/funcref.sgml [hunk 4]
```diff
@@ -5757,6 +5806,44 @@ memory allocated for the driver.
...
+<tag/Availability/cc65
+</descrip>
+</quote>
+
+<sect1>ntohs<label id="ntohs"><p>
+
+<quote>
+<descrip>
+<tag/Function/Swaps byte order in a 16 bit word.
+<tag/Header/<tt/<ref id="stdlib.h" name="stdlib.h">/
```
Comment (oliverschmidt) [inline doc/funcref.sgml:5833]: Missing header adjustment.

File: libsrc/common/ntohl.s [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+;
+; int __fastcall__ ntohl (long val);
+;
+
+.export     _ntohl, _htonl
+.import     popa
+.importzp   tmp1, tmp2, sreg
+
+_htonl := _ntohl
+.code
```
Comment (oliverschmidt) [inline libsrc/common/ntohl.s:12]: The predominant style is to leverage the fact that the assembler defaults to `.code`.

File: libsrc/common/ntohs.s [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+;
+; int __fastcall__ ntohs (int val);
+;
+
+.export     _ntohs, _htons
+.importzp   tmp1
+
+_htons := _ntohs
+
+.code
```
Comment (oliverschmidt) [inline libsrc/common/ntohs.s:12]: The predominant style is to leverage the fact that the assembler defaults to `.code`.

**Comment by colinleroy** [comment]
Thanks, should be better now.

**Comment by kugelfuhr** [comment]
At least for ntohs/htons, it might be an option to use a macro:

```C
#define ntohs(x)                \
    (                           \
        __AX__=(x),             \
        asm("sta tmp1"),        \
        asm("txa"),             \
        asm("ldx tmp1"),        \
        __AX__                  \
    )
```

The code above is 5 bytes. A call is three bytes plus ntohs as a subroutine (6 bytes), making 9 bytes total.

One might also use `__OPT_i__` to either use the macro or call a subroutine.

**Comment by colinleroy** [comment]
Something like that? :)

File: include/arpa/inet.h [hunk 7]
```diff
@@ -31,8 +31,22 @@
 #ifndef _ARPA_INET_H
 #define _ARPA_INET_H
 
+#ifndef __OPT_i__
```
Comment (kugelfuhr) [inline include/arpa/inet.h:34]: `#if (__OPT_i__ < 200)`

`__OPT_i__` contains the argument to `--codesize` or just 200 if `-Oi` was given.

**Comment by colinleroy** [comment]
Thanks!

File: include/arpa/inet.h
[the remark this answers is not in the document]
Comment (oliverschmidt) [reply include/arpa/inet.h:1]: Sorry for the nitpicking... but I was actually referring to the whole thing:
* Number of new lines
* "Code" header
* Comment on the final `#endif`

[the remark this answers is not in the document]
Comment (colinleroy) [reply include/arpa/inet.h:1]: Nitpicking is OK, it makes a code base legible. I'll adapt :)

File: include/arpa/inet.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        asm("ldx tmp1"),        \
+        __AX__                  \
+    )
+#define htons(x) ntohs(x)
+
+#endif
+
+long __fastcall__ ntohl (long val);
+long __fastcall__ htonl (long val);
+
```
Comment (oliverschmidt) [inline include/arpa/inet.h:63]: Two more new lines and we're there...

Comment (colinleroy) [reply include/arpa/inet.h:63]: Right! sorry. Should be good now :)

**Review by oliverschmidt — APPROVED** [review]

**Comment by mrdudz** [comment]
OK :)

**Pull request merged**